WIP: Configure dom0 rpm repo settings via boostrap rpm package.

rocodes opened the pull request:
## Status

Work in progress, pushing in draft mode to discuss approach 
Blocked on release of securedrop-workstation-keyring boostrap rpm (to fpf repos at minimum)

## Description of Changes

Refs #945
Description TK
rpm version check compat todo (for details on rpm constraints, see https://rpm-software-management.github.io/rpm/manual/boolean_dependencies.html )

Changes proposed in this pull request:

## Testing

If you made non-trivial code changes, include a test plan and validated it for this PR.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing pilot instances
2. New installs

## Checklist

### If you have made changes to the provisioning logic

- [ ] All tests (`make test`) pass in `dom0`

### If you have added or removed files

- [ ] I have updated `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

### If documentation is required

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-workstation-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation

rocodes commented:
(It's that the dev keyring is being installed instead of the prod one; just moved it back to in progress to fix up the test that's checking)

rocodes commented:
@deeplow : the existing openqa `install_dev` target installs a prod config.json:

https://github.com/deeplow/openqa-tests-qubesos/blob/8ee37b89929459f0099d9f48461d6468ea0c53d5/tests/securedrop/install_workstation.pm#L103-L105

I'm going to adjust this in a branch and I'll file a ticket here to track.

rocodes commented on tests/test_dom0_rpm_repo.py:
```diff
@@ -1,55 +1,87 @@
-    def test_rpm_repo_public_key(self):
-        with open(self.pubkey_actual) as f:
-            pubkey_actual_contents = f.readlines()
+        with open("tests/files/repo_config.json") as repo_config:
```
It could be in a Python dict too; I was trying to make clearly findable/readable/static (ie the only change is in 2027 or if the key fingerprints change).

legoktm replied:
I would personally prefer a Python dict, just so file reading is one less thing that can go wrong or cause side effects in tests but I won't block over it.

rocodes commented on tests/test_dom0_rpm_repo.py:
```diff
@@ -1,55 +1,87 @@
+
+    def test_dom0_has_keyring_package(self):
+        # Prod keyring is always installed
+        assert self._is_installed("securedrop-workstation-keyring")
```
https://github.com/freedomofpress/securedrop-yum-test/pull/72

rocodes replied:
In openqa, I can go back to setting the environment to "prod" in config.json, and that should pull in the prod keyring. It's a bit weird, but I think allows us to test all the changes separately.

rocodes replied:
no need, we published the package to yum-test nightlies: https://github.com/freedomofpress/securedrop-yum-test/commit/b70cd4282f7b75d33900499343d0677250b1ec31 

I think in general we should try to use one branch on our openqa fork for consistency, and just submit an upstream PR when we're happy with the changes

rocodes commented:
Added a commit to address review feedback, but didn't re-trigger openqa; I'll wait til the package is on yum-test then amend commits + force-push so just the last one has the openqa trigger.

rocodes commented:
Pulling in the dependency on the prod keyring early made things 5% more complicated, I'm putting this briefly back in progress.

rocodes commented:
We introduced a problem when we introduced the DIST substitution in the keyring. I've submitted https://github.com/freedomofpress/securedrop-workstation-keyring/pull/30 to correct it; once it's merged I'll rebuild the keyring package for yum-test and yum-test nightlies, then we should be on track.
